{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Loading via Script"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load and link study, samples, expression, variant, and flow cytometry data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This script is used for uploading and linking the data from public FTP/HTTP. Script uses 'Default template' if nothing else is specified.\n",
      "See https://odm-user-guide.readthedocs.io/en/latest/doc-odm-user-guide/import-data-using-api.html\n",
      "* Make sure you have a Genestack API token first.\n",
      "\u001b[31m* Mandatory! \u001b[0mYou need to provide your token via parameter -t or --token [TOKEN]. \n",
      "\u001b[31m* Mandatory! \u001b[0mDon't forget to provide the link to study file in valid format via -s or --study [URL]\n",
      "  OR(!) you can provide accession of existing study via parameter -sa or --study_accession [ACCESSION].\n",
      "  You're allowed to set only one of them (link to study or accession of existing study) \n",
      "\u001b[31m* Mandatory! \u001b[0mProvide link of samples file in valid format via -sm or --samples [URL]. \n",
      "* Optional. Provide expression data file and expression metadata file \n",
      "  via -e or --expression [URL] and -em or --expression_metadata [URL] respectively. \n",
      "\u001b[32m* Optional. \u001b[0mProvide variant data file and variant metadata file \n",
      "  via -v or --variant [URL] and -vm or --variant_metadata [URL] respectively. \n",
      "\u001b[32m* Optional. \u001b[0mProvide flow cytometry data file and flow cytometry metadata file \n",
      "  via -f or --flow_cytometry [URL] and -fm or --flow_cytometry_metadata [URL] respectively. \n",
      "\u001b[32m* Optional. \u001b[0mIf you need to specify different from Default template, then provide template accession \n",
      "via -tmpl or --template [ACCESSION],  Marked as \"Default\" template will be used, if no specific template accession specified. \n",
      "\u001b[32m* Optional. \u001b[0mYou can set server address via -srv or--server (by default it is https://occam.genestack.com/)\n",
      "\n",
      "\u001b[32m* Note. \u001b[0mAlso you can provide several sample files and several signal files for each samples file.\n",
      "This feature works positionally. This mean you must set signal files parameters after corresponding samples file.\n",
      "\n",
      "If all data files are correct, files will be uploaded and linked successfully. \n",
      "-----------------------------------------------------------------------------\n",
      "\u001b[32mExample 1: \u001b[0mIn this case will be created study with one samples files. \n",
      "For the this samples file will be loaded and linked one variants file, one expression file and one flow cytometry file.\n",
      "python load_and_link_data_for_odm.py --token [token] \\ \n",
      "--study http://data_source/study.csv \\ \n",
      "--samples http://data_source/samples.csv \\ \n",
      "--expression http://data_source/expression.gct \\ \n",
      "--expression_metadata http://data_source/expression_metadata.gct.tsv \\ \n",
      "--variant http://data_source/variations.vcf \\ \n",
      "--variant_metadata http://data_source/variations_meta.vcf.csv \\ \n",
      "--flow_cytometry http://data_source/FACS.facs \\ \n",
      "--flow_cytometry_metadata http://data_source/FACS_meta.facs.csv \\ \n",
      "--template GSF0000000\n",
      "\u001b[32mExample 2: \u001b[0mIn this case will be created study with two samples files. \n",
      "For the first samples file will be loaded and linked two expression files. \n",
      "For the second samples file will be loaded and linked one variants file and one flow cytometry file.\n",
      "python load_and_link_data_for_odm.py --token [token] \\ \n",
      "--study http://data_source/study.csv \\ \n",
      "--samples http://data_source/samples_1.csv \\ \n",
      "--expression http://data_source/expression_1.gct \\ \n",
      "--expression_metadata http://data_source/expression_metadata_1.gct.tsv \\ \n",
      "--expression http://data_source/expression_2.gct \\ \n",
      "--expression_metadata http://data_source/expression_metadata_2.gct.tsv \\ \n",
      "--samples http://data_source/samples_2.csv \\ \n",
      "--variant http://data_source/variations.vcf \\ \n",
      "--variant_metadata http://data_source/variations_meta.vcf.csv \\ \n",
      "--flow_cytometry http://data_source/FACS.facs \\ \n",
      "--flow_cytometry_metadata http://data_source/FACS_meta.facs.csv \\ \n",
      "--template GSF0000000\n",
      "\n",
      "study GSF1047274 was added successfully\n",
      "samples were added successfully\n",
      "expression data were added successfully\n",
      "expression data were added successfully\n",
      "expression data were added successfully\n",
      "flow cytometry data were added successfully\n",
      "variant data were added successfully\n",
      "\u001b[32mExecution is finished!\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!python 'load_and_link_data_for_odm.py' --token '<token>' \\\n",
    "--study https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g.study.tsv \\\n",
    "--samples https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g.samples.tsv \\\n",
    "--expression_metadata https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g_tx.gct.tsv \\\n",
    "--expression https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g_tx.gct \\\n",
    "--expression_metadata https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g_tx_log2.gct.tsv \\\n",
    "--expression https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g_tx_log2.gct \\\n",
    "--expression_metadata https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g_px.gct.tsv \\\n",
    "--expression https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g_px.gct \\\n",
    "--flow_cytometry https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g.facs \\\n",
    "--flow_cytometry_metadata https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g.facs.tsv \\\n",
    "--variant https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g.vcf \\\n",
    "--variant_metadata https://bio-test-data.s3.amazonaws.com/1000gEFV/Test_1000g.vcf.tsv \\\n",
    "--template GSF991585 \\\n",
    "--server https://occam.genestack.com --debug"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set instance and token (to query the uploaded data via API)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import integration_curator, os, statistics, time\n",
    "\n",
    "os.environ['PRED_SPOT_HOST'] = 'occam.genestack.com'\n",
    "os.environ['PRED_SPOT_TOKEN'] = '<token>'\n",
    "os.environ['PRED_SPOT_VERSION'] = 'default-released'\n",
    "\n",
    "omics_api = integration_curator.OmicsQueriesApi()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to get 4 samples: 0 seconds\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genestack:accession</th>\n",
       "      <th>Sample Source ID</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Compound / Compound</th>\n",
       "      <th>Compound / Dose</th>\n",
       "      <th>Compound / Dose Unit</th>\n",
       "      <th>My own attribute</th>\n",
       "      <th>Sample Source</th>\n",
       "      <th>Raw FACS</th>\n",
       "      <th>Raw Variant</th>\n",
       "      <th>Raw Expression</th>\n",
       "      <th>Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSF996650</td>\n",
       "      <td>HG00119</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>M</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1000 Genomes Project</td>\n",
       "      <td>~/Flow/HG00119.facs</td>\n",
       "      <td>~/DNA/HG00119.fq</td>\n",
       "      <td>~/RNA/HG00119.fq</td>\n",
       "      <td>British</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSF996651</td>\n",
       "      <td>HG00121</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>F</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1000 Genomes Project</td>\n",
       "      <td>~/Flow/HG00121.facs</td>\n",
       "      <td>~/DNA/HG00121.fq</td>\n",
       "      <td>~/RNA/HG00121.fq</td>\n",
       "      <td>British</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSF996652</td>\n",
       "      <td>HG00183</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>M</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1000 Genomes Project</td>\n",
       "      <td>~/Flow/HG00183.facs</td>\n",
       "      <td>~/DNA/HG00183.fq</td>\n",
       "      <td>~/RNA/HG00183.fq</td>\n",
       "      <td>Finnish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSF996653</td>\n",
       "      <td>HG00176</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>F</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1000 Genomes Project</td>\n",
       "      <td>~/Flow/HG00176.facs</td>\n",
       "      <td>~/DNA/HG00176.fq</td>\n",
       "      <td>~/RNA/HG00176.fq</td>\n",
       "      <td>Finnish</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  genestack:accession Sample Source ID Organism Disease Tissue   Age Sex  \\\n",
       "0           GSF996650          HG00119     None    None   None  None   M   \n",
       "1           GSF996651          HG00121     None    None   None  None   F   \n",
       "2           GSF996652          HG00183     None    None   None  None   M   \n",
       "3           GSF996653          HG00176     None    None   None  None   F   \n",
       "\n",
       "  Compound / Compound Compound / Dose Compound / Dose Unit My own attribute  \\\n",
       "0                None            None                 None             None   \n",
       "1                None            None                 None             None   \n",
       "2                None            None                 None             None   \n",
       "3                None            None                 None             None   \n",
       "\n",
       "          Sample Source             Raw FACS       Raw Variant  \\\n",
       "0  1000 Genomes Project  ~/Flow/HG00119.facs  ~/DNA/HG00119.fq   \n",
       "1  1000 Genomes Project  ~/Flow/HG00121.facs  ~/DNA/HG00121.fq   \n",
       "2  1000 Genomes Project  ~/Flow/HG00183.facs  ~/DNA/HG00183.fq   \n",
       "3  1000 Genomes Project  ~/Flow/HG00176.facs  ~/DNA/HG00176.fq   \n",
       "\n",
       "     Raw Expression Population  \n",
       "0  ~/RNA/HG00119.fq    British  \n",
       "1  ~/RNA/HG00121.fq    British  \n",
       "2  ~/RNA/HG00183.fq    Finnish  \n",
       "3  ~/RNA/HG00176.fq    Finnish  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "data = omics_api.search_samples(\n",
    "    study_filter='genestack:accession=GSF996648', \n",
    "    sample_filter='\"Population\"=\"British\" OR \"Population\"=\"Finnish\"'\n",
    ")\n",
    "samples = pd.DataFrame.from_dict([item['metadata'] for item in data.data])\n",
    "print('Time to get %s samples: %i seconds\\n' % (samples.shape[0], time.time()-start))\n",
    "\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get and compare genotypes across groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to get 8 genotypes: 0 seconds\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>British AF (AC/AN)</th>\n",
       "      <th>Finnish AF (AC/AN)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Ref / Alt</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs17007017</th>\n",
       "      <th>4:142142653</th>\n",
       "      <th>G / A</th>\n",
       "      <td>0.25 (1/4)</td>\n",
       "      <td>0.75 (3/4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs201966773</th>\n",
       "      <th>4:142142999</th>\n",
       "      <th>T / TC</th>\n",
       "      <td>0.25 (1/4)</td>\n",
       "      <td>1.0 (4/4)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  British AF (AC/AN) Finnish AF (AC/AN)\n",
       "ID          Location    Ref / Alt                                      \n",
       "rs17007017  4:142142653 G / A             0.25 (1/4)         0.75 (3/4)\n",
       "rs201966773 4:142142999 T / TC            0.25 (1/4)          1.0 (4/4)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "data = omics_api.search_variant_data(\n",
    "    study_filter='genestack:accession=GSF996648',\n",
    "    sample_filter='\"Population\"=\"British\" OR \"Population\"=\"Finnish\"',\n",
    "    vx_query='Intervals=4:142142600-142143000',\n",
    "    page_limit=20000\n",
    ")\n",
    "\n",
    "def normalise_genotype(gt): return '1|0' if gt == '0|1' else gt\n",
    "genotypes = pd.DataFrame.from_dict({'genestack:accession': x['relationships']['sample'], \n",
    "                                    'Genotype': normalise_genotype(x['genotype']['GT']),\n",
    "                                    'Location': '%s:%s' % (x['contig'],x['start']),\n",
    "                                    'ID': ', '.join(x['variationId']),\n",
    "                                    'Ref / Alt': x['reference'] + ' / ' + ', '.join(x['alteration'])\n",
    "                                   } \n",
    "                                   for x in data.data)\n",
    "\n",
    "print('Time to get %s genotypes: %i seconds\\n' % (genotypes.shape[0], time.time()-start))\n",
    "\n",
    "samples_genotypes = pd.merge(samples, genotypes)\n",
    "def f(x):\n",
    "    d = {}\n",
    "    for group in x['Population']:\n",
    "        genotypes = x.loc[x['Population'] == group, 'Genotype']\n",
    "        genotypes = '|'.join(genotypes).split('|')\n",
    "        ac = sum([gt == '1' for gt in genotypes])\n",
    "        an = ac + sum([gt == '0' for gt in genotypes])\n",
    "        af = round(ac/an, 2)\n",
    "        d[group+' AF (AC/AN)'] = '%s (%s/%s)' % (af, ac, an)\n",
    "    return pd.Series(d)\n",
    "    \n",
    "samples_genotypes.groupby(['ID', 'Location', 'Ref / Alt']).apply(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get and compare expression values across groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to get 4 expression values: 0 seconds\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>British Median Expression (Q1, Q3)</th>\n",
       "      <th>Finnish Median Expression (Q1, Q3)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000109445</th>\n",
       "      <td>15 (15, 16)</td>\n",
       "      <td>17 (18, 18)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                British Median Expression (Q1, Q3)  \\\n",
       "Gene                                                 \n",
       "ENSG00000109445                        15 (15, 16)   \n",
       "\n",
       "                Finnish Median Expression (Q1, Q3)  \n",
       "Gene                                                \n",
       "ENSG00000109445                        17 (18, 18)  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "data = omics_api.search_expression_data(\n",
    "    study_filter='genestack:accession=GSF996648',\n",
    "    sample_filter='\"Population\"=\"British\" OR \"Population\"=\"Finnish\"',\n",
    "    ex_query = 'Gene=ENSG00000109445 MinValue=0.0',\n",
    "    ex_filter='\"Data Species\"=\"Proteomics (Mass spectrometry)\"'\n",
    ")\n",
    "\n",
    "expressions = pd.DataFrame.from_dict({'genestack:accession': item['relationships']['sample'], \n",
    "                                      'expression': item['expression'],\n",
    "                                     'Gene': item['gene']} for item in data.data)\n",
    "\n",
    "print('Time to get %s expression values: %i seconds\\n' % (expressions.shape[0], time.time()-start))\n",
    "\n",
    "samples_expressions = pd.merge(samples, expressions)\n",
    "def f(x):\n",
    "    d = {}\n",
    "    for group in x['Population']:\n",
    "        exprs = x.loc[x['Population'] == group, 'expression']\n",
    "        quartiles = exprs.quantile([.25, .5, .75]) \n",
    "        d[group+' Median Expression (Q1, Q3)'] = '%i (%i, %i)' % (quartiles.iloc[0], quartiles.iloc[1], quartiles.iloc[2])\n",
    "    return pd.Series(d)\n",
    "    \n",
    "samples_expressions.groupby(['Gene']).apply(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get and compare flow cytometry counts across groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to get 112 flow cytometry cell counts: 0 seconds\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>British Mean Cell Count</th>\n",
       "      <th>Finnish Mean Cell Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CellType</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD45+, live/CD45+, CD3+/CD4</th>\n",
       "      <td>1244</td>\n",
       "      <td>4337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD45+, live/CD45+, CD3+/CD4+/CD4 CM CCR7+ CD45RA+</th>\n",
       "      <td>1157</td>\n",
       "      <td>3104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD45+, live/CD45+, CD3+/CD4+/CD4 CM CCR7+ CD45RA+/CD4 CM</th>\n",
       "      <td>1018</td>\n",
       "      <td>1796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD45+, live/CD45+, CD3+/CD4+/CD4 EM</th>\n",
       "      <td>1403</td>\n",
       "      <td>5023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD45+, live/CD45+, CD3+/CD4+/CD4 EM/CD4 EM1</th>\n",
       "      <td>1255</td>\n",
       "      <td>2605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   British Mean Cell Count  \\\n",
       "CellType                                                                     \n",
       "CD45+, live/CD45+, CD3+/CD4                                           1244   \n",
       "CD45+, live/CD45+, CD3+/CD4+/CD4 CM CCR7+ CD45RA+                     1157   \n",
       "CD45+, live/CD45+, CD3+/CD4+/CD4 CM CCR7+ CD45R...                    1018   \n",
       "CD45+, live/CD45+, CD3+/CD4+/CD4 EM                                   1403   \n",
       "CD45+, live/CD45+, CD3+/CD4+/CD4 EM/CD4 EM1                           1255   \n",
       "\n",
       "                                                   Finnish Mean Cell Count  \n",
       "CellType                                                                    \n",
       "CD45+, live/CD45+, CD3+/CD4                                           4337  \n",
       "CD45+, live/CD45+, CD3+/CD4+/CD4 CM CCR7+ CD45RA+                     3104  \n",
       "CD45+, live/CD45+, CD3+/CD4+/CD4 CM CCR7+ CD45R...                    1796  \n",
       "CD45+, live/CD45+, CD3+/CD4+/CD4 EM                                   5023  \n",
       "CD45+, live/CD45+, CD3+/CD4+/CD4 EM/CD4 EM1                           2605  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "data = omics_api.search_flow_cytometry_data(\n",
    "    study_filter='genestack:accession=GSF996648',\n",
    "    sample_filter='\"Population\"=\"British\" OR \"Population\"=\"Finnish\"',\n",
    "    fx_query='Marker=ENSG00000109445'\n",
    ")\n",
    "\n",
    "cellcounts = pd.DataFrame.from_dict({'genestack:accession': item['relationships']['sample'], \n",
    "                                     'CellType': item['cellPopulation'],\n",
    "                                     'Count': item['expression']} for item in data.data)\n",
    "\n",
    "print('Time to get %s flow cytometry cell counts: %i seconds\\n' % (cellcounts.shape[0], time.time()-start))\n",
    "\n",
    "\n",
    "samples_cells = pd.merge(samples, cellcounts)\n",
    "samples_cells[\"Count\"] = samples_cells[\"Count\"].fillna(0)\n",
    "\n",
    "\n",
    "def f(x):\n",
    "    d = {}\n",
    "    for group in x['Population']:\n",
    "        exprs = x.loc[x['Population'] == group, 'Count']\n",
    "        d[group+' Mean Cell Count'] = '%i' % exprs.mean()\n",
    "    return pd.Series(d)\n",
    "\n",
    "samples_cells.groupby(['CellType']).apply(f).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
