{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using Genestack Omics APIs for eQTL analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Connect to the instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os, time\n",
    "import integration_curator # Genestack client library\n",
    "\n",
    "os.environ['PRED_SPOT_HOST'] = 'occam.genestack.com'\n",
    "os.environ['PRED_SPOT_TOKEN'] = '<your token>'\n",
    "os.environ['PRED_SPOT_VERSION'] = 'default-released'\n",
    "\n",
    "omics_api = integration_curator.OmicsQueriesApi()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Get samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to get 182 samples: 0 seconds\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genestack:accession</th>\n",
       "      <th>Sample Source ID</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Cell Type</th>\n",
       "      <th>Cell Line</th>\n",
       "      <th>Sampling Site</th>\n",
       "      <th>Age</th>\n",
       "      <th>...</th>\n",
       "      <th>Compound Dose</th>\n",
       "      <th>Compound Dose Unit</th>\n",
       "      <th>Raw Data Files</th>\n",
       "      <th>Processed Data Files</th>\n",
       "      <th>Import Source URL</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Health</th>\n",
       "      <th>INDIVIDUALNAME</th>\n",
       "      <th>Sample Source</th>\n",
       "      <th>Species Or Strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSF535900</td>\n",
       "      <td>HG00111</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>HG00111</td>\n",
       "      <td>IGSR</td>\n",
       "      <td>British</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSF535899</td>\n",
       "      <td>HG00110</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>HG00110</td>\n",
       "      <td>IGSR</td>\n",
       "      <td>British</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSF535902</td>\n",
       "      <td>HG00113</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>HG00113</td>\n",
       "      <td>IGSR</td>\n",
       "      <td>British</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSF535901</td>\n",
       "      <td>HG00112</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>HG00112</td>\n",
       "      <td>IGSR</td>\n",
       "      <td>British</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSF535896</td>\n",
       "      <td>HG00106</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>HG00106</td>\n",
       "      <td>IGSR</td>\n",
       "      <td>British</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  genestack:accession Sample Source ID Sample Name Organism Disease Tissue  \\\n",
       "0           GSF535900          HG00111        None     None    None   None   \n",
       "1           GSF535899          HG00110        None     None    None   None   \n",
       "2           GSF535902          HG00113        None     None    None   None   \n",
       "3           GSF535901          HG00112        None     None    None   None   \n",
       "4           GSF535896          HG00106        None     None    None   None   \n",
       "\n",
       "  Cell Type Cell Line Sampling Site   Age  ... Compound Dose  \\\n",
       "0      None      None          None  None  ...          None   \n",
       "1      None      None          None  None  ...          None   \n",
       "2      None      None          None  None  ...          None   \n",
       "3      None      None          None  None  ...          None   \n",
       "4      None      None          None  None  ...          None   \n",
       "\n",
       "  Compound Dose Unit Raw Data Files Processed Data Files Import Source URL  \\\n",
       "0               None           None                 None              None   \n",
       "1               None           None                 None              None   \n",
       "2               None           None                 None              None   \n",
       "3               None           None                 None              None   \n",
       "4               None           None                 None              None   \n",
       "\n",
       "    Sex   Health INDIVIDUALNAME Sample Source Species Or Strain  \n",
       "0  None  Healthy        HG00111          IGSR           British  \n",
       "1  None  Healthy        HG00110          IGSR           British  \n",
       "2  None  Healthy        HG00113          IGSR           British  \n",
       "3  None  Healthy        HG00112          IGSR           British  \n",
       "4  None  Healthy        HG00106          IGSR           British  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_filter = 'genestack:accession=GSF535886'\n",
    "sample_field = 'Species Or Strain'\n",
    "sample_filter = ' OR '.join(['\"%s\"=\"%s\"' % (sample_field, x) \n",
    "                             for x in ['British', 'Finnish']])\n",
    "\n",
    "start = time.time()\n",
    "data = omics_api.search_samples(\n",
    "    study_filter=study_filter, \n",
    "    sample_filter=sample_filter\n",
    ")\n",
    "samples = pd.DataFrame.from_dict([item['metadata'] for item in data.data])\n",
    "print('Time to get %s samples: %i seconds\\n' % (samples.shape[0], time.time()-start))\n",
    "\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Get and compare genotypes across groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to get 724 genotypes: 1 seconds\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>British AF (AC/AN)</th>\n",
       "      <th>Finnish AF (AC/AN)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Ref / Alt</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs10023024</th>\n",
       "      <th>4:142142687</th>\n",
       "      <th>A / G</th>\n",
       "      <td>0.0 (0/176)</td>\n",
       "      <td>0.0 (0/186)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs146362755</th>\n",
       "      <th>4:142142703</th>\n",
       "      <th>A / G</th>\n",
       "      <td>0.01 (2/176)</td>\n",
       "      <td>0.0 (0/186)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs148317497</th>\n",
       "      <th>4:142142719</th>\n",
       "      <th>C / T</th>\n",
       "      <td>0.0 (0/176)</td>\n",
       "      <td>0.0 (0/186)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs17007017</th>\n",
       "      <th>4:142142729</th>\n",
       "      <th>A / G</th>\n",
       "      <td>0.26 (46/176)</td>\n",
       "      <td>0.19 (36/186)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  British AF (AC/AN) Finnish AF (AC/AN)\n",
       "ID          Location    Ref / Alt                                      \n",
       "rs10023024  4:142142687 A / G            0.0 (0/176)        0.0 (0/186)\n",
       "rs146362755 4:142142703 A / G           0.01 (2/176)        0.0 (0/186)\n",
       "rs148317497 4:142142719 C / T            0.0 (0/176)        0.0 (0/186)\n",
       "rs17007017  4:142142729 A / G          0.26 (46/176)      0.19 (36/186)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vx_query = 'Intervals=4:142142600-142143000'\n",
    "# vx_query = 'Gene=ENSG00000109445'\n",
    "\n",
    "start = time.time()\n",
    "data = omics_api.search_variant_data(\n",
    "    study_filter=study_filter,\n",
    "    sample_filter=sample_filter,\n",
    "    vx_query=vx_query,\n",
    "    page_limit=20000\n",
    ")\n",
    "def normalise_genotype(gt): return '1|0' if gt == '0|1' else gt\n",
    "genotypes = pd.DataFrame.from_dict({'genestack:accession': x['relationships']['sample'], \n",
    "                                    'Genotype': normalise_genotype(x['genotype']['GT']),\n",
    "                                    'Location': '%s:%s' % (x['contig'],x['start']),\n",
    "                                    'ID': ', '.join(x['variationId']),\n",
    "                                    'Ref / Alt': x['reference'] + ' / ' + ', '.join(x['alteration'])\n",
    "                                   } \n",
    "                                   for x in data.data)\n",
    "\n",
    "print('Time to get %s genotypes: %i seconds\\n' % (genotypes.shape[0], time.time()-start))\n",
    "\n",
    "samples_genotypes = pd.merge(samples, genotypes)\n",
    "def f(x):\n",
    "    d = {}\n",
    "    for group in x[sample_field]:\n",
    "        genotypes = x.loc[x[sample_field] == group, 'Genotype']\n",
    "        genotypes = '|'.join(genotypes).split('|')\n",
    "        ac = sum([gt == '1' for gt in genotypes])\n",
    "        an = ac + sum([gt == '0' for gt in genotypes])\n",
    "        af = round(ac/an, 2)\n",
    "        d[group+' AF (AC/AN)'] = '%s (%s/%s)' % (af, ac, an)\n",
    "    return pd.Series(d)\n",
    "    \n",
    "samples_genotypes.groupby(['ID', 'Location', 'Ref / Alt']).apply(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Get and compare expression values across groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to get 174 expression values: 0 seconds\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>British Median Expression (Q1, Q3)</th>\n",
       "      <th>Finnish Median Expression (Q1, Q3)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000109445</th>\n",
       "      <td>45 (50, 55)</td>\n",
       "      <td>46 (51, 56)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                British Median Expression (Q1, Q3)  \\\n",
       "Gene                                                 \n",
       "ENSG00000109445                        45 (50, 55)   \n",
       "\n",
       "                Finnish Median Expression (Q1, Q3)  \n",
       "Gene                                                \n",
       "ENSG00000109445                        46 (51, 56)  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = 'ENSG00000109445'\n",
    "ex_query = 'Gene=%s MinValue=0.0' % gene\n",
    "\n",
    "start = time.time()\n",
    "data = omics_api.search_expression_data(\n",
    "    study_filter=study_filter,\n",
    "    sample_filter=sample_filter,\n",
    "    ex_query=ex_query\n",
    ")\n",
    "expressions = pd.DataFrame.from_dict({'genestack:accession': item['relationships']['sample'], \n",
    "                                      'expression': item['expression'],\n",
    "                                     'Gene': item['gene']} for item in data.data)\n",
    "print('Time to get %s expression values: %i seconds\\n' % (expressions.shape[0], time.time()-start))\n",
    "\n",
    "samples_expressions = pd.merge(samples, expressions)\n",
    "def f(x):\n",
    "    d = {}\n",
    "    for group in x[sample_field]:\n",
    "        exprs = x.loc[x[sample_field] == group, 'expression']\n",
    "        quartiles = exprs.quantile([.25, .5, .75])\n",
    "        d[group+' Median Expression (Q1, Q3)'] = '%i (%i, %i)' % (quartiles.iloc[0], quartiles.iloc[1], quartiles.iloc[2])\n",
    "    return pd.Series(d)\n",
    "    \n",
    "samples_expressions.groupby(['Gene']).apply(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
