

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Import Data Using API &mdash; Genestack User Tutorials 1.0 documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Genestack User Tutorials
          

          
            
            <img src="../_static/logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Import Data Using API</a><ul>
<li><a class="reference internal" href="#how-can-i-import-data-to-odm">How can I import data to ODM?</a></li>
<li><a class="reference internal" href="#what-can-i-import">What can I import?</a></li>
<li><a class="reference internal" href="#can-i-capture-the-relationships-between-studies-samples-and-signals">Can I capture the relationships between studies, samples, and signals?</a></li>
<li><a class="reference internal" href="#where-can-i-import-the-data-from">Where can I import the data from?</a></li>
<li><a class="reference internal" href="#step-by-step-example">Step-by-step example</a></li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Genestack User Tutorials</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Import Data Using API</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/doc-odm-user-guide./import-data-using-api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="import-data-using-api">
<h1>Import Data Using API<a class="headerlink" href="#import-data-using-api" title="Permalink to this headline">¶</a></h1>
<div class="section" id="how-can-i-import-data-to-odm">
<h2>How can I import data to ODM?<a class="headerlink" href="#how-can-i-import-data-to-odm" title="Permalink to this headline">¶</a></h2>
<p>On this page, you will find a brief explanation about data import via APIs, followed by a step-by-step example that
you can try. Please note that you need to be a member of the curator group in ODM to be able to import and edit data
in ODM.</p>
</div>
<div class="section" id="what-can-i-import">
<h2>What can I import?<a class="headerlink" href="#what-can-i-import" title="Permalink to this headline">¶</a></h2>
<p>You can import studies, samples, and signals:</p>
<ul class="simple">
<li><p><em>Study</em>: the context of an experiment, such as the aim and statistical design.</p></li>
<li><p><em>Sample</em>: the biological attributes of a sample, such as tissue, disease, and treatment.</p></li>
<li><p><em>Signal</em>: expression, variant, or flow cytometry data. You can import the metadata (e.g. genome version, normalization
method, and the locations of raw/processed data in your storage) together with the processed data (e.g. expression counts, genotypes).</p></li>
</ul>
<p>Once imported, studies, samples, and signal metadata will be queryable and editable from both the User Interface and
APIs, whilst the signal data will only queryable via APIs.</p>
<p>The accepted file formats will be described in the step-by-step example.</p>
</div>
<div class="section" id="can-i-capture-the-relationships-between-studies-samples-and-signals">
<h2>Can I capture the relationships between studies, samples, and signals?<a class="headerlink" href="#can-i-capture-the-relationships-between-studies-samples-and-signals" title="Permalink to this headline">¶</a></h2>
<p>Data import has two stages. First, you create studies, samples, and signals separately. Then, you can link them
together: a study can be linked to multiple samples and a sample can be linked to multi-omics signals. Below are
the data model and how it looks in the User Interface:</p>
<a class="reference internal image-reference" href="../_images/data_model.png"><img alt="../_images/data_model.png" class="align-center" src="../_images/data_model.png" style="width: 515.0px; height: 528.0px;" /></a>
<a class="reference internal image-reference" href="../_images/metainfo_editor.png"><img alt="../_images/metainfo_editor.png" class="align-center" src="../_images/metainfo_editor.png" style="width: 495.0px; height: 314.0px;" /></a>
</div>
<div class="section" id="where-can-i-import-the-data-from">
<h2>Where can I import the data from?<a class="headerlink" href="#where-can-i-import-the-data-from" title="Permalink to this headline">¶</a></h2>
<p>You should provide HTTP/FTP file links (which can include authentication credentials) to a storage that’s accessible
by ODM. File uploads from a local computer will be supported in a future release.</p>
</div>
<div class="section" id="step-by-step-example">
<h2>Step-by-step example<a class="headerlink" href="#step-by-step-example" title="Permalink to this headline">¶</a></h2>
<p>As an example, we will import tiny data subsets from the 1000 Genomes Project, consisting of the following files:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.study.tsv">Test_1000g.study.tsv</a>, a tab-delimited file of the study attributes</p></li>
</ul>
<table class="docutils align-center">
<colgroup>
<col style="width: 30%" />
<col style="width: 45%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Study Source</p></th>
<th class="head"><p>Study Description</p></th>
<th class="head"><p>Target Disease</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1000 Genomes Project</p></td>
<td><p>Subset of 1000 Genomes Project</p></td>
<td><p>Healthy</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><a class="reference external" href="https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.samples.tsv">Test_1000g.samples.tsv</a>, a tab-delimited file of sample attributes.</p></li>
</ul>
<table class="docutils align-center">
<colgroup>
<col style="width: 31%" />
<col style="width: 25%" />
<col style="width: 20%" />
<col style="width: 7%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Study Source</p></th>
<th class="head" colspan="2"><p>Sample Source ID | Species</p></th>
<th class="head" colspan="2"><p>Sex | Population</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1000 Genomes Project</p></td>
<td colspan="2"><p>HG00119      | Homo sapiens</p></td>
<td colspan="2"><p>M  | British</p></td>
</tr>
<tr class="row-odd"><td><p>1001 Genomes Project</p></td>
<td colspan="2"><p>HG00121      | Homo sapiens</p></td>
<td colspan="2"><p>F  | British</p></td>
</tr>
<tr class="row-even"><td><p>1002 Genomes Project</p></td>
<td colspan="2"><p>HG00183      | Homo sapiens</p></td>
<td colspan="2"><p>M  | Finnish</p></td>
</tr>
<tr class="row-odd"><td><p>1003 Genomes Project</p></td>
<td colspan="2"><p>HG00176      | Homo sapiens</p></td>
<td colspan="2"><p>F  | Finnish</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><a class="reference external" href="https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.gct">Test_1000g.gct</a>, a <a class="reference external" href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GCT:_Gene_Cluster_Text_file_format_.28.2A.gct.29">GCT</a> file of expression data from multiple sequencing runs</p></li>
</ul>
<table class="docutils align-center">
<colgroup>
<col style="width: 26%" />
<col style="width: 20%" />
<col style="width: 14%" />
<col style="width: 14%" />
<col style="width: 14%" />
<col style="width: 14%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>HG00119</p></th>
<th class="head"><p>HG00121</p></th>
<th class="head"><p>HG00183</p></th>
<th class="head"><p>HG00176</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ENSG00000077044</p></td>
<td></td>
<td><p>14.7</p></td>
<td><p>16.8</p></td>
<td><p>17.2</p></td>
<td><p>19.5</p></td>
</tr>
<tr class="row-odd"><td><p>ENSG00000085982</p></td>
<td></td>
<td><p>4.2</p></td>
<td><p>7.1</p></td>
<td><p>5.5</p></td>
<td><p>6.8</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><a class="reference external" href="https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.gct.tsv">Test_1000g.gct.tsv</a>, a tab-separated file that describes the expression data</p></li>
</ul>
<table class="docutils align-center">
<colgroup>
<col style="width: 36%" />
<col style="width: 36%" />
<col style="width: 28%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Expression Source</p></th>
<th class="head" colspan="2"><p>Normalization Method | Genome Version</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1000 Genomes Project</p></td>
<td colspan="2"><p>RPKM         | GRCh38.91</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><a class="reference external" href="https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.vcf">Test_1000g.vcf</a>, a <a class="reference external" href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF</a> file of variant data from multiple sequencing runs</p></li>
</ul>
<table class="docutils align-center">
<colgroup>
<col style="width: 7%" />
<col style="width: 12%" />
<col style="width: 12%" />
<col style="width: 4%" />
<col style="width: 6%" />
<col style="width: 5%" />
<col style="width: 7%" />
<col style="width: 8%" />
<col style="width: 7%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>#CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>QUAL</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
<th class="head"><p>FORMAT</p></th>
<th class="head"><p>HG00119</p></th>
<th class="head"><p>HG00121</p></th>
<th class="head"><p>HG00183</p></th>
<th class="head"><p>HG00176</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>233364596</p></td>
<td><p>rs838705</p></td>
<td><p>G</p></td>
<td><p>A</p></td>
<td colspan="2"><p>100  |  PASS</p></td>
<td><p>AF=0.64</p></td>
<td><p>GT</p></td>
<td><p>0|0</p></td>
<td><p>0|1</p></td>
<td><p>1|0</p></td>
<td><p>1|1</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>233385915</p></td>
<td><p>rs201966773</p></td>
<td><p>T</p></td>
<td><p>TTC</p></td>
<td colspan="2"><p>987  |  PASS</p></td>
<td><p>AF=0.86</p></td>
<td><p>GT</p></td>
<td><p>0|0</p></td>
<td><p>0|1</p></td>
<td><p>1|1</p></td>
<td><p>1|1</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><a class="reference external" href="https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.vcf.tsv">Test_1000g.vcf.tsv</a>, a tab-separated file that describes the variant data</p></li>
</ul>
<table class="docutils align-center">
<colgroup>
<col style="width: 58%" />
<col style="width: 42%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Variant Source</p></th>
<th class="head"><p>Genome Version</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1000 Genomes Project</p></td>
<td><p>GRCh38.91</p></td>
</tr>
</tbody>
</table>
<p>We will go through the following steps:</p>
<ol class="arabic simple">
<li><p>Generate an API token</p></li>
<li><p>Import a study</p></li>
<li><p>Import samples</p></li>
<li><p>Link samples to study</p></li>
<li><p>Import and link expression data to samples</p></li>
<li><p>Import and link variant data to samples</p></li>
<li><p>Check that you can query the relationships between objects</p></li>
</ol>
<ol class="arabic simple">
<li><p><strong>Generate an API token</strong></p></li>
</ol>
<p>When using the APIs, you need to provide an API token for authentication.
You can generate this by going to your <a class="reference external" href="https://occam.genestack.com/frontend/endpoint/application/run/genestack/profile">profile</a>, which can be found by clicking your username at the top right corner
of the User Interface, or from the Dashboard</p>
<img alt="../_images/dashboard.png" src="../_images/dashboard.png" />
<p>The API token is permanent — there is no expiration date. However, you can revoke it at any time and have multiple
tokens.</p>
<ol class="arabic simple" start="2">
<li><p><strong>Import a study</strong></p></li>
</ol>
<p>There are specific endpoints to import specific data types, as listed in the <strong>Swagger API documentation</strong>. For studies,
you should go to the <em>studyCurator</em> endpoint, use the <strong>POST</strong> method, and supply the file URL:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Genestack-API-Token: &lt;your API token&gt;&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Content-Type: application/json&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Accept: application/json&quot;</span> <span class="o">-</span><span class="n">X</span> <span class="n">POST</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;{&quot;link&quot;:&quot;https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.study.tsv&quot;}&#39;</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">occam</span><span class="o">.</span><span class="n">genestack</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">frontend</span><span class="o">/</span><span class="n">rs</span><span class="o">/</span><span class="n">genestack</span><span class="o">/</span><span class="n">studyCurator</span><span class="o">/</span><span class="n">v0</span><span class="o">.</span><span class="mi">1</span><span class="o">/</span><span class="n">studies</span>
</pre></div>
</div>
<p>If successful, you should see the accession and metadata of the created study in the response. You should also see some
warnings because some of the required attributes from the default metadata template are missing or incorrect:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
  <span class="s2">&quot;warnings&quot;</span><span class="p">:</span> <span class="p">[</span>
    <span class="s2">&quot;The value for required attribute </span><span class="se">\&quot;</span><span class="s2">Study Source ID</span><span class="se">\&quot;</span><span class="s2"> is not present&quot;</span><span class="p">,</span>
    <span class="s2">&quot;The value for required attribute </span><span class="se">\&quot;</span><span class="s2">Study Type</span><span class="se">\&quot;</span><span class="s2"> is not present&quot;</span>
  <span class="p">],</span>
  <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282799&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Study Source&quot;</span><span class="p">:</span> <span class="s2">&quot;1000 Genomes Project&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Study Source ID&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Study Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Subset of 1000 Genomes Project&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Study Type&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Study Design&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Therapeutic Area&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Contributor&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Contributor Email&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Release Date&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Last Update Date&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Raw Data Files&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Processed Data Files&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Processed Data Matrix Files&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Import Source URL&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Database Source URL&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
    <span class="s2">&quot;Study Title&quot;</span><span class="p">:</span> <span class="s2">&quot;New Study&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Target Disease&quot;</span><span class="p">:</span> <span class="s2">&quot;Healthy&quot;</span>
  <span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
<p>You can also confirm this visually, by going to the <a class="reference external" href="https://occam.genestack.com/frontend/endpoint/application/run/genestack/studybrowser-odm?accessFilter=genestack%3Apersonal">Study Browser</a> and check that a new study has been created,
owned by you:</p>
<img alt="../_images/empty_study.png" src="../_images/empty_study.png" />
<p>To import samples, you should use a different endpoint, <em>sampleCurator</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Genestack-API-Token: &lt;your API token&gt;&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Content-Type: application/json&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Accept: application/json&quot;</span> <span class="o">-</span><span class="n">X</span> <span class="n">POST</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;{&quot;link&quot;:&quot;https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.samples.tsv&quot;}&#39;</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">occam</span><span class="o">.</span><span class="n">genestack</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">frontend</span><span class="o">/</span><span class="n">rs</span><span class="o">/</span><span class="n">genestack</span><span class="o">/</span><span class="n">sampleCurator</span><span class="o">/</span><span class="n">v0</span><span class="o">.</span><span class="mi">1</span><span class="o">/</span><span class="n">samples</span>
</pre></div>
</div>
<p>Similar to the previous step, you should see the accession and metadata of the created samples in the response —
a list of four objects this time, rather than just one object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>[
  {
    &quot;warnings&quot;: [
      &quot;The value \&quot;M\&quot; for attribute \&quot;Sex\&quot; is a synonym in the controlled vocabulary \&quot;GSF000071\&quot;. Did you mean one of [male]?&quot;
    ],
    &quot;data&quot;: {
      &quot;genestack:accession&quot;: &quot;GSF282801&quot;,
      &quot;Sample Source ID&quot;: &quot;HG00119&quot;,
      &quot;Sample Name&quot;: null,
      &quot;Organism&quot;: null,
      &quot;Sex&quot;: &quot;M&quot;,
      &quot;Disease&quot;: null,
      &quot;Cell Line&quot;: null,
      &quot;Cell Type&quot;: null,
      &quot;Tissue&quot;: null,
      &quot;Sampling Site&quot;: null,
      &quot;Age&quot;: null,
      &quot;Compound&quot;: null,
      &quot;Compound Dose&quot;: null,
      &quot;Compound Dose Unit&quot;: null,
      &quot;Raw Data Files&quot;: null,
      &quot;Processed Data Files&quot;: null,
      &quot;Import Source URL&quot;: null,
      &quot;Sample Source&quot;: &quot;1000 Genomes Project&quot;,
      &quot;Population&quot;: &quot;British&quot;,
      &quot;Species&quot;: &quot;Homo sapiens&quot;
    }
  },
… &lt;three more sample objects here&gt; …
]
</pre></div>
</div>
<p>However, you won’t see the samples in the Study Browser yet, because no samples have been linked to the study.</p>
<ol class="arabic simple" start="4">
<li><p><strong>Link samples to study</strong></p></li>
</ol>
<p>You can link samples to study using the integration endpoint, specifying the accessions of the pair of objects to be
linked. The following call will link one of the samples (sample HG00119 with accession GSF282801) to the study
(with accession GSF282799):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Genestack-API-Token: &lt;your API token&gt;&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Content-Type: application/json&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Accept: application/json&quot;</span> <span class="o">-</span><span class="n">X</span> <span class="n">POST</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">occam</span><span class="o">.</span><span class="n">genestack</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">frontend</span><span class="o">/</span><span class="n">rs</span><span class="o">/</span><span class="n">genestack</span><span class="o">/</span><span class="n">integrationCurator</span><span class="o">/</span><span class="n">v0</span><span class="o">.</span><span class="mi">1</span><span class="o">/</span><span class="n">integration</span><span class="o">/</span><span class="n">link</span><span class="o">/</span><span class="n">sample</span><span class="o">/</span><span class="n">GSF282801</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">study</span><span class="o">/</span><span class="n">GSF282799</span>
</pre></div>
</div>
<p>If successful, in the Study Browser you should see (after refreshing the page) that the number of samples next
to your study has changed from ‘-’ to ‘1’:</p>
<img alt="../_images/sample_added.png" src="../_images/sample_added.png" />
<p>You can link the other samples in the same way. In practice, we’d create a script to automate this process, using common sample IDs.</p>
<ol class="arabic simple" start="5">
<li><p><strong>Import and link expression data to samples</strong></p></li>
</ol>
<p>This time, we’re going to import expression data, supplying two files, one for the metadata, and another for the
processed data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Genestack-API-Token: &lt;your API token&gt;&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Content-Type: application/json&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Accept: application/json&quot;</span> <span class="o">-</span><span class="n">X</span> <span class="n">POST</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;{&quot;link&quot;:&quot;https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.gct&quot;, &quot;metadataLink&quot;:&quot;https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.gct.tsv&quot;}&#39;</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">occam</span><span class="o">.</span><span class="n">genestack</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">frontend</span><span class="o">/</span><span class="n">rs</span><span class="o">/</span><span class="n">genestack</span><span class="o">/</span><span class="n">expressionCurator</span><span class="o">/</span><span class="n">v0</span><span class="o">.</span><span class="mi">1</span><span class="o">/</span><span class="n">expression</span><span class="o">/</span><span class="n">gct</span>
</pre></div>
</div>
<p>If successful, you will get four run-level expression objects, corresponding to the four columns in the expression matrix:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
  <span class="s2">&quot;runs&quot;</span><span class="p">:</span> <span class="p">[</span>
    <span class="p">{</span>
      <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282812&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Sample Source ID&quot;</span><span class="p">:</span> <span class="s2">&quot;HG00119&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282813&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Sample Source ID&quot;</span><span class="p">:</span> <span class="s2">&quot;HG00121&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282814&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Sample Source ID&quot;</span><span class="p">:</span> <span class="s2">&quot;HG00183&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282815&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Sample Source ID&quot;</span><span class="p">:</span> <span class="s2">&quot;HG00176&quot;</span>
    <span class="p">}</span>
  <span class="p">],</span>
  <span class="s2">&quot;experiment&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282811&quot;</span><span class="p">,</span>
  <span class="s2">&quot;warnings&quot;</span><span class="p">:</span> <span class="p">[]</span>
<span class="p">}</span>
</pre></div>
</div>
<p>You can then link these run-level expression objects to the corresponding samples. The call below will link
the expression object GSF282812 to the sample HG00119 (that’s been linked to the study in the previous step):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Genestack-API-Token: &lt;your API token&gt;&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Content-Type: application/json&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Accept: application/json&quot;</span> <span class="o">-</span><span class="n">X</span> <span class="n">POST</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">occam</span><span class="o">.</span><span class="n">genestack</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">frontend</span><span class="o">/</span><span class="n">rs</span><span class="o">/</span><span class="n">genestack</span><span class="o">/</span><span class="n">integrationCurator</span><span class="o">/</span><span class="n">v0</span><span class="o">.</span><span class="mi">1</span><span class="o">/</span><span class="n">integration</span><span class="o">/</span><span class="n">link</span><span class="o">/</span><span class="n">expression</span><span class="o">/</span><span class="n">GSF282812</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">sample</span><span class="o">/</span><span class="n">GSF282801</span>
</pre></div>
</div>
<p>If successful, in the Study Browser you should see (after refreshing the page) that the Signals of your study has
changed from ‘-’ to ‘Exp’.</p>
<img alt="../_images/expression_added.png" src="../_images/expression_added.png" />
<ol class="arabic simple" start="6">
<li><p><strong>Import and link variant data to samples</strong></p></li>
</ol>
<p>Let’s repeat the previous step. This time, for variant data, so that the same sample (HG00119) is linked to both
expression and variant data.</p>
<p>To import the variant data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Genestack-API-Token: &lt;your API token&gt;&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Content-Type: application/json&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Accept: application/json&quot;</span> <span class="o">-</span><span class="n">X</span> <span class="n">POST</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;{&quot;link&quot;:&quot;https://s3.amazonaws.com/bio-test-data/Demo/Test_1000g/Test_1000g.vcf&quot;, &quot;metadataLink&quot;:&quot;https://s3.amazonaws.com/bio-test-data/odm/Test_1000g/Test_1000g.vcf.tsv&quot;}&#39;</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">occam</span><span class="o">.</span><span class="n">genestack</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">frontend</span><span class="o">/</span><span class="n">rs</span><span class="o">/</span><span class="n">genestack</span><span class="o">/</span><span class="n">variantCurator</span><span class="o">/</span><span class="n">v0</span><span class="o">.</span><span class="mi">1</span><span class="o">/</span><span class="n">variant</span><span class="o">/</span><span class="n">vcf</span>
</pre></div>
</div>
<p>Which should give you four run-level variant objects, corresponding to the four sample columns in the VCF file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
  <span class="s2">&quot;runs&quot;</span><span class="p">:</span> <span class="p">[</span>
    <span class="p">{</span>
      <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282825&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Sample Source ID&quot;</span><span class="p">:</span> <span class="s2">&quot;HG00119&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282826&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Sample Source ID&quot;</span><span class="p">:</span> <span class="s2">&quot;HG00121&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282827&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Sample Source ID&quot;</span><span class="p">:</span> <span class="s2">&quot;HG00183&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s2">&quot;genestack:accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282828&quot;</span><span class="p">,</span>
      <span class="s2">&quot;Sample Source ID&quot;</span><span class="p">:</span> <span class="s2">&quot;HG00176&quot;</span>
    <span class="p">}</span>
  <span class="p">],</span>
  <span class="s2">&quot;experiment&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282824&quot;</span><span class="p">,</span>
  <span class="s2">&quot;warnings&quot;</span><span class="p">:</span> <span class="p">[]</span>
<span class="p">}</span>
</pre></div>
</div>
<p>To link the variant object GSF282825 with the sample HG00119:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Genestack-API-Token: &lt;your API token&gt;&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Content-Type: application/json&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Accept: application/json&quot;</span> <span class="o">-</span><span class="n">X</span> <span class="n">POST</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">occam</span><span class="o">.</span><span class="n">genestack</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">frontend</span><span class="o">/</span><span class="n">rs</span><span class="o">/</span><span class="n">genestack</span><span class="o">/</span><span class="n">integrationCurator</span><span class="o">/</span><span class="n">v0</span><span class="o">.</span><span class="mi">1</span><span class="o">/</span><span class="n">integration</span><span class="o">/</span><span class="n">link</span><span class="o">/</span><span class="n">variant</span><span class="o">/</span><span class="n">GSF282825</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">sample</span><span class="o">/</span><span class="n">GSF282801</span>
</pre></div>
</div>
<p>Check in the Study Browser that you see both expression and variant signals next to your imported study:</p>
<img alt="../_images/variant_added.png" src="../_images/variant_added.png" />
<ol class="arabic simple" start="7">
<li><p><strong>Check that you can query the relationships between objects</strong></p></li>
</ol>
<p>Once you’ve created and linked the study, sample, expression, and variant objects, you can do integration-aware query via both the User Interface and APIs.</p>
<p>In the User Interface, you should be able to find your imported study using the study, sample, and signal filters:</p>
<img alt="../_images/variant_filter.png" src="../_images/variant_filter.png" />
<p>To do this via APIs, you can use the integration/omics endpoint to filter across studies, samples, and signals,
and retrieve a specific object type. For example, to get the expression of gene ENSG00000077044 from the sample GSF282801:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Genestack-API-Token: &lt;your API token&gt;&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Content-Type: application/json&quot;</span> <span class="o">-</span><span class="n">H</span> <span class="s2">&quot;Accept: application/json&quot;</span> <span class="o">-</span><span class="n">X</span> <span class="n">GET</span> <span class="s2">&quot;https://occam.genestack.com/frontend/rs/genestack/integrationCurator/v0.1/omics/expression/data?sampleFilter=genestack:accession=GSF282801&amp;exQuery=Gene=ENSG00000077044&quot;</span>
</pre></div>
</div>
<p>Which will return:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
  <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span>
    <span class="p">{</span>
      <span class="s2">&quot;itemId&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
      <span class="s2">&quot;metadata&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
      <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="s2">&quot;ENSG00000077044&quot;</span><span class="p">,</span>
      <span class="s2">&quot;expression&quot;</span><span class="p">:</span> <span class="mf">14.74187</span><span class="p">,</span>
      <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
      <span class="s2">&quot;relationships&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="s2">&quot;GSF282801&quot;</span>
      <span class="p">}</span>
    <span class="p">}</span>
  <span class="p">]</span>
<span class="p">}</span>


</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Genestack

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>